### PR TITLE
feat(interview-prep): Converting and Tranfering Rosetta challenges

### PR DESCRIPTION
**Pull request by NitronR**
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Closes #XXXXX

### CODE REVIEW

**Comment by NitronR** [comment]
@scissorsneedfoodtoo This is the next set of problems, can you please review this? I will try to add guide pages for these challenges as @moT01 suggested.

**Review by moT01 — CHANGES_REQUESTED** [review]
That's very strange, when and where did `S-Expressions` get deleted from the meta? It helps if you can give a little explanation in the description when you find something like that @NitronR. I didn't review the challenges themselves yet.

The challenge files should all be changed to end with `.english.md` and not just `.md`

I created guide articles for these...
https://www.freecodecamp.org/forum/t/freecodecamp-challenge-guide-search-a-list-of-records/385315
https://www.freecodecamp.org/forum/t/freecodecamp-challenge-guide-self-referential-sequence/385317
https://www.freecodecamp.org/forum/t/freecodecamp-challenge-guide-semiprime/385318
https://www.freecodecamp.org/forum/t/freecodecamp-challenge-guide-set-consolidation/385319
https://www.freecodecamp.org/forum/t/freecodecamp-challenge-guide-set-of-real-numbers/385322
https://www.freecodecamp.org/forum/t/freecodecamp-challenge-guide-sha-1/385326
https://www.freecodecamp.org/forum/t/freecodecamp-challenge-guide-sha-256/385328

You just need to add the forumTopicId in the frontmatter.

**Note to self: add these solutions to the guide articles when this gets merged.**

**Comment by NitronR** [comment]
@moT01 Thank you for adding the guide pages. I have added  the forumTopicIds and renamed the files. 
I corrected the position of `S-Expressions` so that it can be consistent with [Rosetta Code ordering](http://www.rosettacode.org/wiki/Category:Programming_Tasks). I am sorry, I forgot to mention it in the description.

**Comment by moT01** [comment]
Thanks @NitronR, 

> I corrected the position of S-Expressions

Ahh, I guess I missed the removal of it in the diff. I was looking through the file history and couldn't find where it got deleted.

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/search-a-list-of-records.english.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/search-a-list-of-records.english.md:50]: const lst = [

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/search-a-list-of-records.english.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/search-a-list-of-records.english.md:61]: Not a big deal, but we might consider moving `lst` below `searchCity` so the function that campers will work on is at the top of the editor.

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/search-a-list-of-records.english.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/search-a-list-of-records.english.md:92]: Not the biggest deal since users won't see this, but we could consider updating this solution to use ES6.

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/self-referential-sequence.english.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/self-referential-sequence.english.md:19]: No code blocks here, but the spacing can match the original problem. Let's also remove one unnecessary link and go to the secure Rosetta Code page:

```
<section id='description'>
There are several ways to generate a self-referential sequence. One very common one (the <a href="https://rosettacode.org/wiki/Look-and-say sequence" target="_blank">Look-and-say sequence</a>) is to start with a positive integer, then generate the next term by concatenating enumerated groups of adjacent alike digits:

<pre>0, 10, 1110, 3110, 132110, 1113122110, 311311222110 ...</pre>

The terms generated grow in length geometrically and never converge.

Another way to generate a self-referential sequence is to summarize the previous term.

Count how many of each alike digit there is, then concatenate the sum and digit for each of the sorted enumerated digits. Note that the first five terms are the same as for the previous sequence.

<pre>0, 10, 1110, 3110, 132110, 13123110, 23124110 ...</pre>

Sort the digits largest to smallest. Do not include counts of digits that do not appear in the previous term.

Depending on the seed value, series generated this way always either converge to a stable value or to a short cyclical pattern. (For our purposes, I'll use converge to mean an element matches a previously seen element.) The sequence shown, with a seed value of 0, converges to a stable value of 1433223110 after 11 iterations. The seed value that converges most quickly is 22. It goes stable after the first element. (The next element is 22, which has been seen before.)
</section>
```

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/semiprime.english.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/semiprime.english.md:18]: Write a function that returns true if a number is semiprime, or false if it is not.

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/semiprime.english.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/semiprime.english.md:12]:
````
Semiprime numbers are natural numbers that are products of exactly two (possibly equal) <a href="https://rosettacode.org/wiki/prime_number" target="_blank">prime numbers</a>.
<pre>1679  =  23 x 73</pre>
````

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/set-consolidation.english.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/set-consolidation.english.md:22]:
````
<section id='description'>
Given two sets of items then if any item is common to any set then the result of applying <i>consolidation</i> to those sets is a set of sets whose contents is:

<ul>
  <li>The two input sets if no common item exists between the two input sets of items.</li>
  <li>The single set that is the union of the two input sets if they share a common item.</li>
</ul>

Given N sets of items where N > 2 then the result is the same as repeatedly replacing all combinations of two sets by their consolidation until no further consolidation between set pairs is possible. If N < 2 then consolidation has no strict meaning and the input can be returned.

Here are some examples:

<strong>Example 1:</strong>

<span style="margin-left: 2em;">Given the two sets `{A,B}` and `{C,D}` then there is no common element between the sets and the result is the same as the input.</span>

<strong>Example 2:</strong>

<span style="margin-left: 2em;">Given the two sets `{A,B}` and `{B,D}` then there is a common element `B` between the sets and the result is the single set `{B,D,A}`.  (Note that order of items in a set is immaterial: `{A,B,D}` is the same as `{B,D,A}` and `{D,A,B}`, etc).</span>

<strong>Example 3:</strong>

<span style="margin-left: 2em;">Given the three sets `{A,B}` and `{C,D}` and `{D,B}` then there is no common element between the sets `{A,B}` and `{C,D}` but the sets `{A,B}` and `{D,B}` do share a common element that consolidates to produce the result `{B,D,A}`. On examining this result with the remaining set, `{C,D}`, they share a common element and so consolidate to the final output of the single set `{A,B,C,D}`</span>

<strong>Example 4:</strong>

<span style="margin-left: 2em;">The consolidation of the five sets:</span>

<span style="margin-left: 4em;">`{H,I,K}`, `{A,B}`, `{C,D}`, `{D,B}`, and `{F,G,H}`</span>

<span style="margin-left: 2em;">Is the two sets:</span>

<span style="margin-left: 4em;">`{A, C, B, D}`, and `{G, F, I, H, K}`</span>
</section>
````

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/set-of-real-numbers.english.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/set-of-real-numbers.english.md:32]:
````
<section id='description'>
All real numbers form the uncountable set ℝ.  Among its subsets, relatively simple are the convex sets, each expressed as a range between two real numbers <i>a</i> and <i>b</i> where <i>a</i> ≤ <i>b</i>.  There are actually four cases for the meaning of "between", depending on open or closed boundary:

<ul>
  <li>[<i>a</i>, <i>b</i>]: {<i>x</i> | <i>a</i> ≤ <i>x</i> and <i>x</i> ≤ <i>b</i> }</li>
  <li>(<i>a</i>, <i>b</i>): {<i>x</i> | <i>a</i> < <i>x</i> and <i>x</i> < <i>b</i> }</li>
  <li>[<i>a</i>, <i>b</i>): {<i>x</i> | <i>a</i> ≤ <i>x</i> and <i>x</i> < <i>b</i> }</li>
  <li>(<i>a</i>, <i>b</i>]: {<i>x</i> | <i>a</i> < <i>x</i> and <i>x</i> ≤ <i>b</i> }</li>
</ul>

Note that if <i>a</i> = <i>b</i>, of the four only [<i>a</i>, <i>a</i>] would be non-empty.

<strong>Task</strong>

<ul>
  <li>Devise a way to represent any set of real numbers, for the definition of "any" in the implementation notes below.</li>
  <li>Provide methods for these common set operations (<i>x</i> is a real number; <i>A</i> and <i>B</i> are sets):</li>
  <ul>
    <li>
      <i>x</i> ∈ <i>A</i>: determine if <i>x</i> is an element of <i>A</i><br>
      example: 1 is in [1, 2), while 2, 3, ... are not.
    </li>
    <li>
      <i>A</i> ∪ <i>B</i>: union of <i>A</i> and <i>B</i>, i.e. {<i>x</i> | <i>x</i> ∈ <i>A</i> or <i>x</i> ∈ <i>B</i>}<br>
      example: [0, 2) ∪ (1, 3) = [0, 3); [0, 1) ∪ (2, 3] = well, [0, 1) ∪ (2, 3]
    </li>
    <li>
      <i>A</i> ∩ <i>B</i>: intersection of <i>A</i> and <i>B</i>, i.e. {<i>x</i> | <i>x</i> ∈ <i>A</i> and <i>x</i> ∈ <i>B</i>}<br>
      example: [0, 2) ∩ (1, 3) = (1, 2); [0, 1) ∩ (2, 3] = empty set
    </li>
    <li>
      <i>A</i> - <i>B</i>: difference between <i>A</i> and <i>B</i>, also written as <i>A</i> \ <i>B</i>, i.e. {<i>x</i> | <i>x</i> ∈ <i>A</i> and <i>x</i> ∉ <i>B</i>}<br>
      example: [0, 2) − (1, 3) = [0, 1]
    </li>
  </ul>
</ul>
</section>
````

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/set-of-real-numbers.english.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/set-of-real-numbers.english.md:38]:
````
<section id='instructions'>

Write a function that takes 2 objects, a string and an array as parameters. The objects represents the set and have attributes: `low`, `high` and `rangeType`. 

The `rangeType` can have values 0, 1, 2 and 3 for `CLOSED`, `BOTH_OPEN`, `LEFT_OPEN` and `RIGHT_OPEN`, respectively. The function should implement a set using this information. 

The string represents the operation to be performed on the sets. It can be: `"union"`, `"intersect"` and `"subtract"` (difference). 

After performing the operation, the function should check if the values in the array are present in the resultant set and store a corresponding boolean value to an array. The function should return this array.

</section>
````

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/sha-1.english.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/sha-1.english.md:15]: Spacing here seems alright. Let's just go to the secure version of the page:

```
A US government standard, <a href="https://rosettacode.org/wiki/SHA-1/FIPS-180-1" target="_blank">FIPS 180-1</a>, defines SHA-1.
```

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/sha-256.english.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/sha-256.english.md:11]:
```
<b><a href="https://en.wikipedia.org/wiki/SHA-256" target="_blank">SHA-256</a></b> is the recommended stronger alternative to <a href="https://rosettacode.org/wiki/SHA-1" target="_blank">SHA-1</a>.
```

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/sha-256.english.md
Comment (scissorsneedfoodtoo) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/sha-256.english.md:17]: Write a function that takes a string as a parameter and returns its SHA-256 digest.

**Comment by scissorsneedfoodtoo** [comment]
@moT01, awesome, thank you for creating Guide articles for all of those challenges.

@NitronR, I noticed a few things and will leave some review comments.

**Comment by scissorsneedfoodtoo** [comment]
@thecodingaviator, could you take a look at this when you get a chance?

**Comment by thecodingaviator** [comment]
@NitronR drop a reply in after you've made the above changes, and I'll have a look!

**Comment by NitronR** [comment]
@scissorsneedfoodtoo Thanks for the review. I have done the changes.
@thecodingaviator Can you please take a look?

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/self-referential-sequence.english.md
Comment (thecodingaviator) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/self-referential-sequence.english.md:11]: We might want to add what a self-referential sequence is

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/self-referential-sequence.english.md
Comment (thecodingaviator) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/self-referential-sequence.english.md:25]:
```suggestion
Depending on the seed value, series generated this way always either converge to a stable value or to a short cyclical pattern. (For our purposes, I'll use converge to mean an element matches a previously seen element.) The sequence shown, with a seed value of 0, converges to a stable value of 1433223110 after 11 iterations. The seed value that converges most quickly is 22. It goes stable after the first element. (The next element is 22, which has been seen before.)
```
Uses self-referencing language `I'll` and such. Please revise this

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/semiprime.english.md
Comment (thecodingaviator) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/semiprime.english.md:18]:
```suggestion
Write a function that returns true if a number is semiprime, and false if it is not.
```

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/sha-1.english.md
Comment (thecodingaviator) [inline curriculum/challenges/english/08-coding-interview-prep/rosetta-code/sha-1.english.md:22]: I think we should add how SHA-1 is generated and such relevant information.

**Review by thecodingaviator — CHANGES_REQUESTED** [review]

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/self-referential-sequence.english.md
Comment (NitronR) [reply curriculum/challenges/english/08-coding-interview-prep/rosetta-code/self-referential-sequence.english.md:11]: I did not find a formal definition but mostly its just the essence of the words "self-referential sequence" which we can describe. We can add this line: A self-referential sequence is a sequence whose elements are generated by using its current elements.
What do you think?

Comment (thecodingaviator) [reply curriculum/challenges/english/08-coding-interview-prep/rosetta-code/self-referential-sequence.english.md:11]: I guess rather than coming up with one of our own, we should leave it to the user to understand in that case?

Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/08-coding-interview-prep/rosetta-code/self-referential-sequence.english.md:11]: @thecodingaviator, I think it's okay to let learners look up the definition or go to the Look-and-say sequence to get a better understanding of what a self-referential sequence is. Also if they read on they get some concrete examples.

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/self-referential-sequence.english.md
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/08-coding-interview-prep/rosetta-code/self-referential-sequence.english.md:25]: This is a tricky once since it's in the original Rosetta Code description. Generally we want to stick closely to the original description, but here it does make sense to modify it.

How about we change the sentence slightly to remove the word "I'll":

"(For our purposes, converge meaning when an element matches a previously seen element.)"

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/semiprime.english.md
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/08-coding-interview-prep/rosetta-code/semiprime.english.md:18]: This could go either way. It seems like our current challenge use both "and" or "or" for this type of sentence.

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/sha-1.english.md
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/08-coding-interview-prep/rosetta-code/sha-1.english.md:22]: I don't think this is necessary in this case. If we do, the description will be quite long.

The included link toward the bottom of the description breaks things down well. Alternatively, we could also include a link to Wikipedia at the top:

`<a href="https://en.wikipedia.org/wiki/SHA-1" target="_blank"><b>SHA-1</b></a> or <b>SHA1</b> is a one-way hash function; it computes a 160-bit message digest.`

Comment (thecodingaviator) [reply curriculum/challenges/english/08-coding-interview-prep/rosetta-code/sha-1.english.md:22]: Didn't see the link before. That works!

File: curriculum/challenges/english/08-coding-interview-prep/rosetta-code/self-referential-sequence.english.md
Comment (thecodingaviator) [reply curriculum/challenges/english/08-coding-interview-prep/rosetta-code/self-referential-sequence.english.md:25]: That sounds great!

**Review by thecodingaviator — DISMISSED** [review]

**Comment by NitronR** [comment]
@scissorsneedfoodtoo @thecodingaviator Thanks for helping out with this!

**Review by moT01 — CHANGES_REQUESTED** [review]
Can we also add the new `isHidden: false` flag at the top of these.

**Comment by moT01** [comment]
I have blocked this for now. See the [pinned code freeze issue](https://github.com/freeCodeCamp/freeCodeCamp/issues/38308)

**Comment by raisedadead** [comment]
Limitations removed on Coding Interview Prep Challenges.

**Comment by scissorsneedfoodtoo** [comment]
@NitronR, hope you don't mind, but I went in and added the `isHidden` flags for you.

Thanks again for all of your hard work porting these challenges over. Looking forward to your next PR :+1: :+1:

**Comment by NitronR** [comment]
@scissorsneedfoodtoo Thank you for the help!

**Review by moT01 — CHANGES_REQUESTED** [review]
I made a few suggestions @NitronR. Nice job on these 👍 Could you complete the checklist in the PR template please?

File: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/self-referential-sequence.english.md
Comment (NitronR) [reply curriculum/challenges/english/10-coding-interview-prep/rosetta-code/self-referential-sequence.english.md:32]: @moT01 I think the second method is the one in which the series converges and hence it is the focus of the task. For the first method, it is stated: "The terms generated grow in length geometrically and never converge". As it is an infinite sequence, it cannot have a finite answer. I also think that we can make the instructions more clear.
```
Write a function that takes the seed value as parameter, generate a  self referential sequence until it converges and return it as an array.
```
This would imply to use the second method as the first one do not converge.

File: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/self-referential-sequence.english.md
Comment (NitronR) [reply curriculum/challenges/english/10-coding-interview-prep/rosetta-code/self-referential-sequence.english.md:45]: Makes sense. I will update the test cases and the solution.

File: curriculum/challenges/english/10-coding-interview-prep/rosetta-code/self-referential-sequence.english.md
Comment (scissorsneedfoodtoo) [reply curriculum/challenges/english/10-coding-interview-prep/rosetta-code/self-referential-sequence.english.md:32]: @NitronR, thanks for improving these instructions. Your proposal sounds much clearer. I just want to suggest a couple of small changes:

`Write a function that takes the seed value as parameter, generates a self referential sequence until it converges, and returns it as an array.`

**Comment by NitronR** [comment]
@moT01 @scissorsneedfoodtoo I have made the changes. Can you please check?

**Review by moT01 — APPROVED** [review]
LGTM @NitronR 🎉 Thanks for all the work on this 🎉

**Comment by scissorsneedfoodtoo** [comment]
Thanks for all your hard work on these ports @NitronR and @moT01 for your review. Everything LGTM :+1: :+1:

**Pull request merged**
